### PR TITLE
[SPARK-44396][Connect] Direct Arrow Deserialization

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds direct arrow to user object deserialization to the Spark Connect Scala Client.

### Why are the changes needed?
We want to decouple the scala client from catalyst. We need a way to encode user object from and to arrrow.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added tests to `ArrowEncoderSuite`.

### CODE REVIEW

**Comment by hvanhovell** [comment]
Scala 2.13 compilation is still broken, and I need to add a couple of tests.

**Review by zhenlineo — APPROVED** [review]

**Review by zhenlineo — APPROVED** [review]
I left some minor doc suggestions to help understand the code. They can be addressed later.

Can you also update the PR description with the purpose of the PR?
Is this mostly for a better structure of how to decode arrow to our native types e.g. Map, Array etc.? Do we expect performance gain too from the change? Dose our own serdes make it faster to obtain the user objects too? If so, can you add some doc in the code too?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkResult.scala [hunk 1]
```diff
@@ -76,61 +71,63 @@ private[sql] class SparkResult[T](
...
-              if (stopOnFirstNonEmptyResponse) {
-                return true
-              }
-            }
+        val reader = new MessageIterator(ipcStreamBytes.newInput(), allocator)
+        if (arrowSchema == null) {
+          arrowSchema = reader.schema
+          stop |= stopOnArrowSchema
+        } else if (arrowSchema != reader.schema) {
+          // Uh oh...
```
[the remark this answers is not in the document]
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkResult.scala:95]: Done. The only issue is that this incredibly difficult to test.

**Comment by hvanhovell** [comment]
Merging this. Test failure is unrelated.

**Pull request closed without merging**